### PR TITLE
Wrap object in structure that contains all needed informations

### PR DESCRIPTION
**Pull request by AurelienFT**
## Short explanation

The goal of this PR is reduce the computation of the object ID and the serialize version of an object that is sent through the network.

Some schemas for the example of an operation:
Deserialization
Serialization

## Detailed explanation 

The `Signed` structure has been renamed to `Wrapped` and now contains `id`, `serialized_data`, `creator_public_key`, `creator_address`. This structure have a serializer and a deserializer

The serializer encode only the public key and the serialized data.
The deserializer deduces the others fields thanks to this data.

This PR implement new serializer and deserializer for Block, BlockHeader, Operation, Endorsement because they are needed for the Serializer and Deserializer of WrappedXXX structure.

The serialization and deserialization can be customized. This is needed for the `Block` because we don't want to encode to much information as a lot are shared his `header` field.

Everywhere in the code (except at the creation) we must have `WrappedXXXX` structure because it prevent to re-compute a lot of data.

To create this type of structures you need to use the constructor that is implemented for all objects that are passed in the network (currently Block,BlockHeader,Operation,Endorsement).

This PR also implement officially the RFC #2690 even if it was already the behavior in the current codebase.

This PR simplify a lot of code since the serialized version is directly in the object (especially in storage).

This PR also removes all the constraints on all serializers on the code because the verification should be made only on deserialization.

### CODE REVIEW

File: massa-models/src/block.rs [hunk 1]
```diff
@@ -583,28 +583,27 @@ mod test {
...
-                    .1,
+                    .unwrap(),
                     Wrapped::new_wrapped(
                         Endorsement {
-                            sender_public_key: public_key,
                             slot: Slot::new(4, 0),
                             index: 3,
                             endorsed_block: BlockId(Hash::compute_from("blk2".as_bytes())),
                         },
+                        EndorsementSerializer::new(),
```
Comment (damip) [inline massa-models/src/block.rs:600]: could we share the EndorsementSerializer created above instead of creating a new one ?

**Comment by damip** [comment]
looks OK for now :)

**Comment by AurelienFT** [comment]
> looks OK for now :)

Thank you for taking a look :)

**Comment by AurelienFT** [comment]
@massalabs/core-team This PR is ready for the review. I have written explanations in the PR description above. Also I ping all of you because this PR modify a lot of things across the code and each of you knows well each part of the code.

If you have any question to not hesitate.

**Review by gterzian — APPROVED** [review]
Looks good to me, I'm just wondering if we can hide the instantiations of the (de-) serializers inside the implementation of `Wrapped`.

**Review by Eitu33 — APPROVED** [review]
Very nice changes 👍

File: massa-models/src/block.rs [hunk 2]
```diff
@@ -106,34 +124,194 @@ impl BlockId {
...
+}
+
+impl BlockDeserializer {
+    /// Creates a new `BlockDeserializer`
+    pub fn new() -> Self {
+        BlockDeserializer {
+            header_deserializer: WrappedDeserializer::new(BlockHeaderDeserializer::new()),
+            operation_deserializer: WrappedDeserializer::new(OperationDeserializer::new()),
+            u32_deserializer: U32VarIntDeserializer::new(
+                Included(0),
```
Comment (damip) [inline massa-models/src/block.rs:251]: why use 2 bounds ? why not a Range ? https://doc.rust-lang.org/std/ops/struct.Range.html

Is it to ensure that we always specify bounds ?

File: massa-models/src/endorsement.rs [hunk 3]
```diff
@@ -119,128 +123,131 @@ pub struct Endorsement {
...
+impl EndorsementDeserializer {
+    /// Creates a new `EndorsementDeserializer`
+    pub fn new(endorsement_count: u32) -> Self {
+        #[cfg(feature = "sandbox")]
+        let thread_count = *THREAD_COUNT;
+        #[cfg(not(feature = "sandbox"))]
+        let thread_count = THREAD_COUNT;
+        EndorsementDeserializer {
+            slot_deserializer: SlotDeserializer::new(
+                (Included(1), Included(u64::MAX)),
```
Comment (damip) [inline massa-models/src/endorsement.rs:178]: I think we should start from 0 included here

File: massa-models/src/block.rs [hunk 4]
```diff
@@ -207,36 +386,178 @@ impl std::fmt::Display for Block {
...
+    /// Creates a new `BlockHeaderDeserializer`
+    pub fn new() -> Self {
+        #[cfg(feature = "sandbox")]
+        let thread_count = *THREAD_COUNT;
+        #[cfg(not(feature = "sandbox"))]
+        let thread_count = THREAD_COUNT;
+        Self {
+            slot_deserializer: SlotDeserializer::new(
+                (Included(0), Included(u64::MAX)),
+                (Included(0), Included(thread_count)),
```
Comment (damip) [inline massa-models/src/block.rs:486]: thread_count is excluded

File: massa-models/src/endorsement.rs [hunk 5]
```diff
@@ -119,128 +123,131 @@ pub struct Endorsement {
...
+                (Included(0), Included(thread_count)),
```
Comment (damip) [inline massa-models/src/endorsement.rs:179]: thread count is excluded

**Review by damip — CHANGES_REQUESTED** [review]
Beautiful and extensive work !
some (security-critical) nits to correct  :)

File: massa-models/src/endorsement.rs (hunk 3 above)
Comment (AurelienFT) [reply massa-models/src/endorsement.rs:178]: Ok I will make the change but the old code had : 
```rust
if slot.period == 0 {
            return Err(ModelsError::DeserializeError(
                "the target period of an endorsement cannot be 0".into(),
            ));
}
```

File: massa-models/src/block.rs (hunk 4 above)
Comment (AurelienFT) [reply massa-models/src/block.rs]: Done.

File: massa-models/src/endorsement.rs (hunk 5 above)
Comment (AurelienFT) [reply massa-models/src/endorsement.rs]: Done.

**Review by damip — APPROVED** [review]
Amazing work

**Comment by AurelienFT** [comment]
bors merge

**Pull request merged**